src/provider/provider_tracking.c

```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
     return ret;
 }
 
+static umf_result_t trackingAllocationSplit(void *provider, void *ptr,
+                                            size_t totalSize,
+                                            size_t firstSize) {
+    umf_result_t ret = UMF_RESULT_ERROR_UNKNOWN;
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *splitValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
+    splitValue->pool = p->pool;
+    splitValue->size = firstSize;
+
+    int r = util_mutex_lock(p->hTracker->splitMergeMutex);
+    if (r) {
+        goto err_lock;
+    }
+
+    tracker_value_t *value =
+        (tracker_value_t *)critnib_get(p->hTracker->map, (uintptr_t)ptr);
+    if (!value) {
+        fprintf(stderr, "tracking split: no such value\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+    if (value->size != totalSize) {
+        fprintf(stderr, "tracking split: %zu != %zu\n", value->size, totalSize);
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+
+    ret = umfMemoryProviderAllocationSplit(p->hUpstream, ptr, totalSize,
+                                           firstSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr,
+                "tracking split: umfMemoryProviderAllocationSplit failed\n");
+        goto err;
+    }
+
+    void *highPtr = (void *)(((uintptr_t)ptr) + firstSize);
+    size_t rightSize = totalSize - firstSize;
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
```

Comment:
1) `have duplicate` -> `have to duplicate` ?
2) `highPtr, rightValue` ? - maybe `highPtr` and `highValue` ?